msg-wire/src/compression/mod.rs

```diff
@@ -0,0 +1,69 @@
+use bytes::Bytes;
+use flate2::{read::GzDecoder, write::GzEncoder, Compression};
+use std::io::{self, Read, Write};
+
+/// This trait is used to implement message-level compression algorithms for payloads.
+/// On outgoing messages, the payload is compressed before being sent using the `compress` method.
+/// On incoming messages, the inverse happens using the `decompress` method.
+pub trait Compressor: Send + Sync + Unpin + 'static {
+    /// Compresses a byte slice payload into a `Bytes` object.
+    fn compress(&self, data: &[u8]) -> Result<Bytes, io::Error>;
+
+    /// Decompresses a compressed byte slice into a `Bytes` object.
+    fn decompress(&self, data: &[u8]) -> Result<Bytes, io::Error>;
+}
+
+/// A compressor that uses the gzip algorithm.
+pub struct GzipCompressor {
+    level: u32,
+}
+
+impl GzipCompressor {
+    /// Creates a new gzip compressor with the given compression level (0-9).
+    pub fn new(level: u32) -> Self {
+        Self { level }
+    }
```

Comment:
can you add an assertion for the compression range check? 
e.g. assert!(level < 10)